## Part 1: Explore the Data

Import the data and use Pandas to learn more about the dataset.

In [1]:
import pandas as pd

df = pd.read_csv('Resources/client_dataset.csv')
df.head()

first      last                    job               phone  \
0     Donald   Harding           Immunologist  793-904-7725x39308   
1    Tiffany     Myers        Music therapist    201.442.4543x942   
2    Shannon    Watson           Immunologist   687.737.9424x8503   
3     Nathan     Baker  Accounting technician    827-788-8123x012   
4  Christina  Schwartz           Chiropractor        265-829-3643   

                                 email  client_id  order_id  order_date  \
0     harding.donald.7185@sullivan.com      58515   8953482  2023-04-28   
1    myers.t.6537@ferguson-johnson.net      37609   8069089  2023-05-19   
2                swatson8146@payne.net      57113   1902144  2023-01-29   
3               bakernathan@benson.com      46554   9031802  2023-04-25   
4  christinaschwartz9252@mcconnell.com      92089   1322274  2023-05-28   

   order_week  order_year         item_id     category         subcategory  \
0          17        2023  EUD29711-63-6U        decor            wall art   
1          20        2023  XDA18116-89-4A  consumables                pens   
2           4        2023  ABE59463-05-7E     software  project management   
3          17        2023  ZMM00836-65-0C  consumables                pens   
4          21        2023  BZX55559-12-3X  consumables                misc   

   unit_price  unit_cost  unit_weight  qty  line_number  
0     1096.80     762.71         7.50  105            1  
1       24.95      15.09         1.49   21            0  
2       13.52       7.86         1.68   39            6  
3       36.42      24.85         1.23   29            3  
4      195.10     108.17        46.43   20            1

In [3]:
# View the column names in the data
print(df.columns)

Index(['first', 'last', 'job', 'phone', 'email', 'client_id', 'order_id',
       'order_date', 'order_week', 'order_year', 'item_id', 'category',
       'subcategory', 'unit_price', 'unit_cost', 'unit_weight', 'qty',
       'line_number'],
      dtype='object')


In [4]:
# Use the describe function to gather some basic statistics
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number  
count  54639.000000  54639.000000  5.463900e+04  54639.000000  
mean      99.446073      5.004116  5.702646e+02      2.979667  
std      133.164267      5.326599  1.879552e+04      2.436320  
min        0.010000      0.000000  0.000000e+00      0.000000  
25%       14.840000      1.450000  3.200000e+01      1.000000  
50%       49.890000      3.240000  6.800000e+01      3.000000  
75%      125.570000      6.890000  1.700000e+02      5.000000  
max      846.270000     46.430000  3.958244e+06      9.000000

In [6]:
# Use this space to do any additional research
# and familiarize yourself with the data.
df.dtypes
# Get value counts for order_id to see frequency of orders
print("Order ID Value Counts:")
print(df['order_id'].value_counts())

# Get basic info about the dataframe
print("\nDataframe Info:")
print(df.info())

# Check for any missing values
print("\nMissing Values:")
print(df.isnull().sum())

# Get unique values in key columns
print("\nUnique Values in Key Columns:")
print(f"Number of unique clients: {df['client_id'].nunique()}")
print(f"Number of unique orders: {df['order_id'].nunique()}")
print(f"Number of unique items: {df['item_id'].nunique()}")

# Get value distributions for important columns
print("\nValue Distributions:")
print("\nCategory Distribution:")
print(df['category'].value_counts())
print("\nSubcategory Distribution:") 
print(df['subcategory'].value_counts())

Order ID Value Counts:
order_id
4397975    10
4330331    10
5673069    10
1095932    10
4070245    10
           ..
1631129     1
2604731     1
3487378     1
1407286     1
6032865     1
Name: count, Length: 10000, dtype: int64

Dataframe Info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 54639 entries, 0 to 54638
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   first        54639 non-null  object 
 1   last         54639 non-null  object 
 2   job          54639 non-null  object 
 3   phone        54639 non-null  object 
 4   email        54639 non-null  object 
 5   client_id    54639 non-null  int64  
 6   order_id     54639 non-null  int64  
 7   order_date   54639 non-null  object 
 8   order_week   54639 non-null  int64  
 9   order_year   54639 non-null  int64  
 10  item_id      54639 non-null  object 
 11  category     54639 non-null  object 
 12  subcategory  54639 non-null  object 
 13  unit_price  

In [8]:
# What three item categories had the most entries?
category_counts = df['category'].value_counts()
print("\nTop 3 Categories by Number of Entries:")
print(category_counts.head(3))

#~Count non-null entries.
non_null_counts = df.count()

#~Select the three columns with the most entries.
top_3 = non_null_counts.nlargest(3)
print(top_3)


Top 3 Categories by Number of Entries:
category
consumables    23538
furniture      11915
software        8400
Name: count, dtype: int64
first    54639
last     54639
job      54639
dtype: int64


In [9]:
# For the category with the most entries, which subcategory had the most entries?
sub_entries = top_3.nlargest(1)
print(sub_entries)

first    54639
dtype: int64


In [10]:
# Which five clients had the most entries in the data?
# Get value counts of client_id to find most frequent clients
client_counts = df['client_id'].value_counts()
print("Top 5 Clients by Number of Orders:")
print(client_counts.head())

#~order the counts based on the number of instances of the clients order data.
most_entries = df['client_id'].value_counts()

#~print the top 5 entries with the .head command.
most_entries.head()

Top 5 Clients by Number of Orders:
client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64


client_id
33615    220
66037    211
46820    209
38378    207
24741    207
Name: count, dtype: int64

In [11]:
# Store the client ids of those top 5 clients in a list.
top_5_clients = list(df['client_id'].value_counts().head(5).index)
top_5_clients


[33615, 66037, 46820, 38378, 24741]

In [12]:
# How many total units (the qty column) did the client with the most entries order?
# Get the client ID with the most entries
top_client_id = df['client_id'].value_counts().idxmax()

# Sum up all quantities ordered by this client
total_units = df[df['client_id'] == top_client_id]['qty'].sum()

# Print the result
print(f"Client {top_client_id} ordered a total of {total_units} units")


Client 33615 ordered a total of 64313 units


## Part 2: Transform the Data
Do we know that this client spent the more money than client 66037? If not, how would we find out? Transform the data using the steps below to prepare it for analysis.

In [13]:
# Create a column that calculates the subtotal for each line using the unit_price and the qty
df['subtotal'] = df['unit_price'] * df['qty']

print(df[['unit_price', 'qty', 'subtotal']])

       unit_price   qty   subtotal
0         1096.80   105  115164.00
1           24.95    21     523.95
2           13.52    39     527.28
3           36.42    29    1056.18
4          195.10    20    3902.00
...           ...   ...        ...
54634       83.13    33    2743.29
54635      206.59    47    9709.73
54636       65.66   475   31188.50
54637        1.48   112     165.76
54638        3.01  1031    3103.31

[54639 rows x 3 columns]


In [14]:
# Create a column for shipping price.
# Assume a shipping price of $7 per pound for orders over 50 pounds and $10 per pound for items 50 pounds or under.
df['shipping_weight'] = df['unit_weight'] * df['qty']

df['shipping_price'] = df.apply(
    lambda row: row['shipping_weight'] * 7 if row['shipping_weight'] > 50 
    else row['shipping_weight'] * 10, axis=1
    )

print(df[['unit_weight', 'qty', 'shipping_weight', 'shipping_price']])

       unit_weight   qty  shipping_weight  shipping_price
0             7.50   105           787.50         5512.50
1             1.49    21            31.29          312.90
2             1.68    39            65.52          458.64
3             1.23    29            35.67          356.70
4            46.43    20           928.60         6500.20
...            ...   ...              ...             ...
54634         2.25    33            74.25          519.75
54635        11.70    47           549.90         3849.30
54636         4.16   475          1976.00        13832.00
54637        18.04   112          2020.48        14143.36
54638         2.07  1031          2134.17        14939.19

[54639 rows x 4 columns]


In [16]:
# First, calculate the tax
TAX_RATE = 0.0925
df['tax'] = (df['subtotal'] + df['shipping_price']) * TAX_RATE

# Then calculate the line total
df['line_total'] = df['subtotal'] + df['shipping_price'] + df['tax']

# Finally calculate the line cost and profit
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']
df['profit'] = df['line_total'] - df['line_cost']

In [17]:
# Create a column for the cost of each line using unit cost, qty, and
# shipping price (assume the shipping cost is exactly what is charged to the client).
df['line_cost'] = (df['unit_cost'] * df['qty']) + df['shipping_price']

#define line total.
df['line_total'] = df['subtotal'] + df['shipping_price'] + df['tax']

#return the line total.
df[['unit_weight', 'qty', 'shipping_weight', 'shipping_price', 'subtotal', 'tax', 'line_total']]
df.describe()

client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number      subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667  7.710286e+04   
std      133.164267      5.326599  1.879552e+04      2.436320  2.881926e+06   
min        0.010000      0.000000  0.000000e+00      0.000000  0.000000e+00   
25%       14.840000      1.450000  3.200000e+01      1.000000  1.132950e+03   
50%       49.890000      3.240000  6.800000e+01      3.000000  4.629440e+03   
75%      125.570000      6.890000  1.700000e+02      5.000000  1.736913e+04   
max      846.270000     46.430000  3.958244e+06      9.000000  5.844743e+08   

       shipping_weight  shipping_price     line_cost           tax  \
count     5.463900e+04    5.463900e+04  5.463900e+04  5.463900e+04   
mean      2.487416e+03    1.742510e+04  7.387972e+04  8.743836e+03   
std       5.559430e+04    3.891595e+05  2.133294e+06  2.782624e+05   
min       0.000000e+00    0.000000e+00  0.000000e+00  0.000000e+00   
25%       7.033000e+01    4.966500e+02  2.206360e+03  2.444849e+02   
50%       2.254000e+02    1.577800e+03  6.266040e+03  7.109152e+02   
75%       7.184400e+02    5.029080e+03  1.946590e+04  2.239986e+03   
max       9.009056e+06    6.306339e+07  3.987535e+08  5.429454e+07   

         line_total        profit  
count  5.463900e+04  5.463900e+04  
mean   1.032718e+05  2.939207e+04  
std    3.286505e+06  1.195291e+06  
min    0.000000e+00  0.000000e+00  
25%    2.887565e+03  5.597773e+02  
50%    8.396485e+03  1.894838e+03  
75%    2.645605e+04  6.674207e+03  
max    6.412625e+08  2.425090e+08

In [18]:
# Create a column for the profit of each line using line cost and line price
df['profit'] = df['line_total'] - df['shipping_price'] - df['tax']

#Return the result.
print(df['profit'])
df.describe()

0        115164.00
1           523.95
2           527.28
3          1056.18
4          3902.00
           ...    
54634      2743.29
54635      9709.73
54636     31188.50
54637       165.76
54638      3103.31
Name: profit, Length: 54639, dtype: float64


client_id      order_id    order_week    order_year    unit_price  \
count  54639.000000  5.463900e+04  54639.000000  54639.000000  54639.000000   
mean   54837.869416  5.470190e+06     11.359139   2022.993064    136.267207   
std    25487.438231  2.599807e+06      7.023499      0.082997    183.873135   
min    10033.000000  1.000886e+06      1.000000   2022.000000      0.010000   
25%    33593.000000  3.196372e+06      6.000000   2023.000000     20.800000   
50%    53305.000000  5.496966e+06     11.000000   2023.000000     68.310000   
75%    78498.000000  7.733869e+06     17.000000   2023.000000    173.160000   
max    99984.000000  9.998480e+06     52.000000   2023.000000   1396.230000   

          unit_cost   unit_weight           qty   line_number      subtotal  \
count  54639.000000  54639.000000  5.463900e+04  54639.000000  5.463900e+04   
mean      99.446073      5.004116  5.702646e+02      2.979667  7.710286e+04   
std      133.164267      5.326599  1.879552e+04      2.436320  2.881926e+06   
min        0.010000      0.000000  0.000000e+00      0.000000  0.000000e+00   
25%       14.840000      1.450000  3.200000e+01      1.000000  1.132950e+03   
50%       49.890000      3.240000  6.800000e+01      3.000000  4.629440e+03   
75%      125.570000      6.890000  1.700000e+02      5.000000  1.736913e+04   
max      846.270000     46.430000  3.958244e+06      9.000000  5.844743e+08   

       shipping_weight  shipping_price     line_cost           tax  \
count     5.463900e+04    5.463900e+04  5.463900e+04  5.463900e+04   
mean      2.487416e+03    1.742510e+04  7.387972e+04  8.743836e+03   
std       5.559430e+04    3.891595e+05  2.133294e+06  2.782624e+05   
min       0.000000e+00    0.000000e+00  0.000000e+00  0.000000e+00   
25%       7.033000e+01    4.966500e+02  2.206360e+03  2.444849e+02   
50%       2.254000e+02    1.577800e+03  6.266040e+03  7.109152e+02   
75%       7.184400e+02    5.029080e+03  1.946590e+04  2.239986e+03   
max       9.009056e+06    6.306339e+07  3.987535e+08  5.429454e+07   

         line_total        profit  
count  5.463900e+04  5.463900e+04  
mean   1.032718e+05  7.710286e+04  
std    3.286505e+06  2.881926e+06  
min    0.000000e+00  0.000000e+00  
25%    2.887565e+03  1.132950e+03  
50%    8.396485e+03  4.629440e+03  
75%    2.645605e+04  1.736913e+04  
max    6.412625e+08  5.844743e+08

## Part 3: Confirm your work
You have email receipts showing that the total prices for 3 orders. Confirm that your calculations match the receipts. Remember, each order has multiple lines.

Order ID 2742071 had a total price of \$152,811.89

Order ID 2173913 had a total price of \$162,388.71

Order ID 6128929 had a total price of \$923,441.25


In [20]:
# Check your work using the totals above
# Filter for the specific order IDs we want to verify
orders_to_check = [2742071, 2173913, 6128929]
order_totals = df[df['order_id'].isin(orders_to_check)]

# Group by order_id and calculate totals
order_totals_df = order_totals.groupby('order_id').agg({
    'first': 'first',
    'last': 'first',
    'client_id': 'first',
    'item_id': lambda x: ','.join(x),
    'unit_weight': 'sum',
    'qty': 'sum', 
    'line_number': 'max',
    'subtotal': 'sum',
    'shipping_weight': 'sum',
    'shipping_price': 'sum',
    'line_cost': 'sum',
    'tax': 'sum',
    'line_total': 'sum',
    'profit': 'sum'
})

# Compare calculated totals with receipt values
print("Comparing calculated totals with receipts:")
print(f"Order 2742071: Calculated ${order_totals_df.loc[2742071, 'line_total']:.2f} vs Receipt $152,811.89")
print(f"Order 2173913: Calculated ${order_totals_df.loc[2173913, 'line_total']:.2f} vs Receipt $162,388.71") 
print(f"Order 6128929: Calculated ${order_totals_df.loc[6128929, 'line_total']:.2f} vs Receipt $923,441.25")

# Display full results
order_totals_df


Comparing calculated totals with receipts:
Order 2742071: Calculated $152811.90 vs Receipt $152,811.89
Order 2173913: Calculated $162388.72 vs Receipt $162,388.71
Order 6128929: Calculated $923441.24 vs Receipt $923,441.25


first     last  client_id  \
order_id                                 
2173913     Darius  Stevens      22122   
2742071   Kimberly    Blair      78447   
6128929    Dorothy     Hall      88257   

                                                    item_id  unit_weight  \
order_id                                                                   
2173913   CDB25978-21-3E,XED42806-14-3M,CXD77584-10-1E,Z...        25.11   
2742071   ACU90518-83-2B,BMD72525-90-2X,DEE52027-40-3U,A...        20.48   
6128929   DUX14130-38-6U,XDX02308-74-0X,UUX03397-27-4C,E...        45.06   

           qty  line_number   subtotal  shipping_weight  shipping_price  \
order_id                                                                  
2173913   1221            6  109440.76          5576.57        39198.80   
2742071    439            5  127799.33          1691.98        12074.26   
6128929   4594            9  688564.27         22364.97       156690.87   

          line_cost           tax     line_total     profit  
order_id                                                     
2173913   130500.17  13749.159300  162388.719300  109440.76  
2742071   124058.01  12938.307075  152811.897075  127799.33  
6128929   625740.71  78186.100450  923441.240450  688564.27

## Part 4: Summarize and Analyze
Use the new columns with confirmed values to find the following information.

In [22]:
# How much did each of the top 5 clients by quantity spend? Check your work from Part 1 for client ids.
def top_5(client_id, column): 
    client_df = df.loc[df['client_id'] == client_id, column]
    return round(client_df.sum(), 2)

for x in top_5_clients:
    print(f"{x}: ${top_5(x, 'line_total')}")


33615: $8377308.57
66037: $10259514.8
46820: $9743794.32
38378: $12906550.88
24741: $82268891.98


In [24]:
# Create a summary DataFrame showing the totals for the top 5 clients with the following information:
# total units purchased, total shipping price, total revenue, and total profit.
top_5_data = df[df['client_id'].isin(top_5_clients)]

summary_data = top_5_data.groupby('client_id').agg({
    'qty': 'sum',
    'shipping_price': 'sum', 
    'subtotal': 'sum',
    'unit_cost': 'sum',
    'profit': 'sum'
}).reset_index()
summary_data

client_id     qty  shipping_price     subtotal  unit_cost       profit
0      24741  239862      5126448.37  70176885.25   17900.04  70176885.25
1      33615   64313      1828984.89   5839032.11   21716.88   5839032.11
2      38378   73667      3429455.40   8384321.15   21702.21   8384321.15
3      46820   75768      1601448.84   7317356.03   18140.06   7317356.03
4      66037   43018      1395151.85   7995708.38   21508.98   7995708.38

In [25]:
summary_data.describe()

client_id            qty  shipping_price      subtotal   unit_cost  \
count      5.00000       5.000000    5.000000e+00  5.000000e+00      5.0000   
mean   41918.20000   99325.600000    2.676298e+06  1.994266e+07  20193.6340   
std    15670.64535   79624.481118    1.587849e+06  2.809854e+07   1987.7094   
min    24741.00000   43018.000000    1.395152e+06  5.839032e+06  17900.0400   
25%    33615.00000   64313.000000    1.601449e+06  7.317356e+06  18140.0600   
50%    38378.00000   73667.000000    1.828985e+06  7.995708e+06  21508.9800   
75%    46820.00000   75768.000000    3.429455e+06  8.384321e+06  21702.2100   
max    66037.00000  239862.000000    5.126448e+06  7.017689e+07  21716.8800   

             profit  
count  5.000000e+00  
mean   1.994266e+07  
std    2.809854e+07  
min    5.839032e+06  
25%    7.317356e+06  
50%    7.995708e+06  
75%    8.384321e+06  
max    7.017689e+07

In [26]:
summary_data.columns

Index(['client_id', 'qty', 'shipping_price', 'subtotal', 'unit_cost',
       'profit'],
      dtype='object')

In [27]:
# Format the data and rename the columns to names suitable for presentation.
summary_data = summary_data.rename(columns={
                                            'client_id' : 'Client ID',
                                            'qty' : 'Quantity'})
# Define the money columns.
money_columns = ['shipping_price','subtotal','unit_cost','profit']


# Define a function that converts a dollar amount to millions.
def currency_format_millions(x):
    return x / 1000000

# Apply the currency_format_millions function to only the money columns.
for column in money_columns:
    summary_data[column] = summary_data[column].apply(currency_format_millions)


# Rename the columns to reflect the change in the money format. 
summary_data = summary_data.rename(columns ={
                                            'shipping_price' : 'Shipping(millions)',
                                            'subtotal' : 'Total Revenue(millions)',
                                            'unit_cost' : 'Total Cost(millions)',
                                            'profit' : 'Total Profit(millions)'})

print(summary_data)

   Client ID  Quantity  Shipping(millions)  Total Revenue(millions)  \
0      24741    239862            5.126448                70.176885   
1      33615     64313            1.828985                 5.839032   
2      38378     73667            3.429455                 8.384321   
3      46820     75768            1.601449                 7.317356   
4      66037     43018            1.395152                 7.995708   

   Total Cost(millions)  Total Profit(millions)  
0              0.017900               70.176885  
1              0.021717                5.839032  
2              0.021702                8.384321  
3              0.018140                7.317356  
4              0.021509                7.995708  


In [28]:
# Sort the updated data by "Total Profit (millions)" form highest to lowest and assign the sort to a new DatFrame.
sorted_profit = summary_data.sort_values(by='Total Profit(millions)', ascending=False)
sorted_profit

Client ID  Quantity  Shipping(millions)  Total Revenue(millions)  \
0      24741    239862            5.126448                70.176885   
2      38378     73667            3.429455                 8.384321   
4      66037     43018            1.395152                 7.995708   
3      46820     75768            1.601449                 7.317356   
1      33615     64313            1.828985                 5.839032   

   Total Cost(millions)  Total Profit(millions)  
0              0.017900               70.176885  
2              0.021702                8.384321  
4              0.021509                7.995708  
3              0.018140                7.317356  
1              0.021717                5.839032